File: src/med_exam_toolkit/stats.py
"""题目统计分析"""
from __future__ import annotations
from collections import Counter
from med_exam_toolkit.models import Question
import unicodedata

DIFFICULTY_LABELS = {
    "easy": "简单 (≥80%)",
    "medium": "中等 (60-80%)",
    "hard": "较难 (40-60%)",
    "extreme": "困难 (<40%)",
    "unknown": "未知 (无正确率)",
}

DIFFICULTY_ORDER = ["easy", "medium", "hard", "extreme", "unknown"]

def _parse_rate(raw: str) -> float | None:
    if not raw or not raw.strip():
        return None
    s = raw.strip().rstrip("%")
    try:
        v = float(s)
        return v if 0 <= v <= 100 else None
    except ValueError:
        return None

def _display_width(s: str) -> int:
    """计算字符串在终端的显示宽度"""
    return sum(2 if unicodedata.east_asian_width(c) in ("F", "W") else 1 for c in s)

def _pad_right(s: str, width: int) -> str:
    """按显示宽度右补空格"""
    return s + " " * (width - _display_width(s))

def _classify_difficulty(q: Question) -> str:
    rates = []
    for sq in q.sub_questions:
        r = _parse_rate(sq.rate)
        if r is not None:
            rates.append(r)
    if not rates:
        return "unknown"
    avg = sum(rates) / len(rates)
    if avg >= 80:
        return "easy"
    if avg >= 60:
        return "medium"
    if avg >= 40:
        return "hard"
    return "extreme"


def summarize(questions: list[Question], full: bool = False) -> dict:
    """生成统计摘要, full=True 时章节/题库不截断"""
    by_mode = Counter()
    by_unit = Counter()
    by_pkg = Counter()
    by_cls = Counter()
    by_difficulty = Counter()
    low_rate_questions = []
    total_subquestions = 0

    for q in questions:
        by_mode[q.mode] += 1
        by_unit[q.unit] += 1
        by_pkg[q.pkg] += 1
        by_cls[q.cls] += 1
        by_difficulty[_classify_difficulty(q)] += 1
        total_subquestions += len(q.sub_questions)

        for sq in q.sub_questions:
            if sq.rate:
                try:
                    rate_val = float(sq.rate.replace("%", "").strip())
                    if rate_val < 50:
                        low_rate_questions.append({
                            "text": sq.text[:60],
                            "rate": sq.rate,
                            "answer": sq.answer,
                            "unit": q.unit,
                            "mode": q.mode,
                        })
                except ValueError:
                    pass

    unit_limit = None if full else 20

    # 按 DIFFICULTY_ORDER 排序
    difficulty_sorted = {
        k: by_difficulty.get(k, 0)
        for k in DIFFICULTY_ORDER
        if by_difficulty.get(k, 0) > 0
    }

    return {
        "total": len(questions),
        "total_subquestions": total_subquestions,
        "by_mode": dict(by_mode.most_common()),
        "by_unit": dict(by_unit.most_common(unit_limit)),
        "by_pkg": dict(by_pkg.most_common()),
        "by_cls": dict(by_cls.most_common()),
        "by_difficulty": difficulty_sorted,
        "unit_total": len(by_unit),
        "low_rate_count": len(low_rate_questions),
        "low_rate_top10": sorted(
            low_rate_questions,
            key=lambda x: float(x["rate"].replace("%", "").strip()),
        )[:10],
        "full": full,
    }


def print_summary(questions: list[Question], full: bool = False) -> None:
    """打印统计摘要到终端"""
    s = summarize(questions, full=full)
    total = s["total"] or 1
    print(f"\n{'='*50}")
    print(f"📊 题目统计")
    print(f"{'='*50}")
    print(f"总题数: {s['total']} 道大题, {s['total_subquestions']} 道小题")

    def _print_section(title: str, data: dict, show_bar: bool = True, show_pct: bool = True):
        print(f"\n{title}:")
        if not data:
            print("  (无数据)")
            return
        # 自动计算标签列宽度
        labels = {k: (k if k.strip() else "未知") for k in data}
        col_width = max(_display_width(v) for v in labels.values()) + 2
        max_count = max(data.values())
        for key, count in data.items():
            label = labels[key]
            padded = _pad_right(label, col_width)
            pct =  f"({count / total * 100:>5.1f}%)" if show_pct else ""
            bar = " " + "■" * round(count / max_count * 20) if show_bar else ""
            print(f"  {padded} {count:>5d} {pct}{bar}")

    _print_section("按题型", s["by_mode"])

    difficulty_labeled = {
        DIFFICULTY_LABELS.get(k, k): v for k, v in s["by_difficulty"].items()
    }
    _print_section("按难度", difficulty_labeled)

    _print_section("按来源", s["by_pkg"])
    _print_section("按题库", s["by_cls"], show_bar=False)

    unit_items = list(s["by_unit"].items())
    if full:
        _print_section(f"按章节 (共 {s['unit_total']} 个)", s["by_unit"], show_bar=False, show_pct=False)
    else:
        top10 = dict(unit_items[:10])
        _print_section(f"按章节 (Top 10 / 共 {s['unit_total']} 个)", top10, show_bar=False, show_pct=False)
        if s["unit_total"] > 10:
            print(f"  ... 还有 {s['unit_total'] - 10} 个章节")

    if s["low_rate_count"]:
        print(f"\n⚠️  正确率 < 50% 的题目: {s['low_rate_count']} 道")
        for item in s["low_rate_top10"]:
            print(f"  [{item['rate']}] {item['text']}...")

    print(f"{'='*50}\n")

File: internal/stats/stats.go
// Package stats provides question-bank statistics and terminal-formatted summaries.
package stats

import (
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
	"unicode/utf8"

	"github.com/zyh001/med-exam-kit/internal/models"
)

// DifficultyLabels maps internal keys to display labels.
var DifficultyLabels = map[string]string{
	"easy":    "简单 (≥80%)",
	"medium":  "中等 (60-80%)",
	"hard":    "较难 (40-60%)",
	"extreme": "困难 (<40%)",
	"unknown": "未知 (无正确率)",
}

var difficultyOrder = []string{"easy", "medium", "hard", "extreme", "unknown"}

// Summary holds the computed statistics.
type Summary struct {
	Total        int
	TotalSubQ    int
	ByMode       []KV
	ByUnit       []KV
	ByPkg        []KV
	ByCls        []KV
	ByDifficulty []KV
	UnitTotal    int
	LowRateCount int
	LowRateTop10 []LowRateItem
	Full         bool
}

type KV struct {
	Key   string
	Count int
}

// LowRateItem represents a question with a low correct rate.
type LowRateItem struct {
	Text   string
	Rate   string
	Answer string
	Unit   string
	Mode   string
}

func parseRate(raw string) (float64, bool) {
	s := strings.TrimSpace(raw)
	if s == "" {
		return 0, false
	}
	s = strings.TrimSuffix(s, "%")
	v, err := strconv.ParseFloat(s, 64)
	if err != nil || v < 0 || v > 100 {
		return 0, false
	}
	return v, true
}

func classifyDifficulty(q *models.Question) string {
	var rates []float64
	for _, sq := range q.SubQuestions {
		if r, ok := parseRate(sq.Rate); ok {
			rates = append(rates, r)
		}
	}
	if len(rates) == 0 {
		return "unknown"
	}
	sum := 0.0
	for _, r := range rates {
		sum += r
	}
	avg := sum / float64(len(rates))
	switch {
	case avg >= 80:
		return "easy"
	case avg >= 60:
		return "medium"
	case avg >= 40:
		return "hard"
	default:
		return "extreme"
	}
}

// Summarize computes statistics for a list of questions.
func Summarize(questions []*models.Question, full bool) Summary {
	byMode := map[string]int{}
	byUnit := map[string]int{}
	byPkg := map[string]int{}
	byCls := map[string]int{}
	byDiff := map[string]int{}
	var lowRate []LowRateItem
	totalSQ := 0

	for _, q := range questions {
		byMode[q.Mode]++
		byUnit[q.Unit]++
		byPkg[q.Pkg]++
		byCls[q.Cls]++
		byDiff[classifyDifficulty(q)]++
		totalSQ += len(q.SubQuestions)

		for _, sq := range q.SubQuestions {
			if r, ok := parseRate(sq.Rate); ok && r < 50 {
				text := sq.Text
				if len([]rune(text)) > 60 {
					text = string([]rune(text)[:60])
				}
				lowRate = append(lowRate, LowRateItem{
					Text: text, Rate: sq.Rate, Answer: sq.Answer,
					Unit: q.Unit, Mode: q.Mode,
				})
			}
		}
	}

	// Sort low rate by rate ascending
	sort.Slice(lowRate, func(i, j int) bool {
		ri, _ := parseRate(lowRate[i].Rate)
		rj, _ := parseRate(lowRate[j].Rate)
		return ri < rj
	})
	top10 := lowRate
	if len(top10) > 10 {
		top10 = top10[:10]
	}

	// Build difficulty in order
	var diffKV []KV
	for _, k := range difficultyOrder {
		if v := byDiff[k]; v > 0 {
			diffKV = append(diffKV, KV{k, v})
		}
	}

	unitLimit := -1
	if !full {
		unitLimit = 20
	}

	return Summary{
		Total:        len(questions),
		TotalSubQ:    totalSQ,
		ByMode:       sortedKV(byMode, -1),
		ByUnit:       sortedKV(byUnit, unitLimit),
		ByPkg:        sortedKV(byPkg, -1),
		ByCls:        sortedKV(byCls, -1),
		ByDifficulty: diffKV,
		UnitTotal:    len(byUnit),
		LowRateCount: len(lowRate),
		LowRateTop10: top10,
		Full:         full,
	}
}

func sortedKV(m map[string]int, limit int) []KV {
	kvs := make([]KV, 0, len(m))
	for k, v := range m {
		kvs = append(kvs, KV{k, v})
	}
	sort.Slice(kvs, func(i, j int) bool {
		return kvs[i].Count > kvs[j].Count
	})
	if limit > 0 && len(kvs) > limit {
		kvs = kvs[:limit]
	}
	return kvs
}

// PrintSummary prints a formatted summary to stdout.
func PrintSummary(questions []*models.Question, full bool) {
	s := Summarize(questions, full)
	total := s.Total
	if total == 0 {
		total = 1
	}

	fmt.Printf("\n%s\n", strings.Repeat("=", 50))
	fmt.Println("📊 题目统计")
	fmt.Println(strings.Repeat("=", 50))
	fmt.Printf("总题数: %d 道大题, %d 道小题\n", s.Total, s.TotalSubQ)

	printSection := func(title string, data []KV, showBar, showPct bool) {
		fmt.Printf("\n%s:\n", title)
		if len(data) == 0 {
			fmt.Println("  (无数据)")
			return
		}
		colWidth := 4
		for _, item := range data {
			label := item.Key
			if strings.TrimSpace(label) == "" {
				label = "未知"
			}
			w := displayWidth(label)
			if w+2 > colWidth {
				colWidth = w + 2
			}
		}
		maxCount := 0
		for _, item := range data {
			if item.Count > maxCount {
				maxCount = item.Count
			}
		}
		for _, item := range data {
			label := item.Key
			if strings.TrimSpace(label) == "" {
				label = "未知"
			}
			padded := padRight(label, colWidth)
			pct := ""
			if showPct {
				pct = fmt.Sprintf("(%5.1f%%)", float64(item.Count)/float64(total)*100)
			}
			bar := ""
			if showBar && maxCount > 0 {
				barLen := int(math.Round(float64(item.Count) / float64(maxCount) * 20))
				bar = " " + strings.Repeat("■", barLen)
			}
			fmt.Printf("  %s %5d %s%s\n", padded, item.Count, pct, bar)
		}
	}

	printSection("按题型", s.ByMode, true, true)

	// Convert difficulty keys to labels
	diffLabeled := make([]KV, len(s.ByDifficulty))
	for i, d := range s.ByDifficulty {
		label := DifficultyLabels[d.Key]
		if label == "" {
			label = d.Key
		}
		diffLabeled[i] = KV{label, d.Count}
	}
	printSection("按难度", diffLabeled, true, true)
	printSection("按来源", s.ByPkg, true, true)
	printSection("按题库", s.ByCls, false, false)

	if full {
		printSection(fmt.Sprintf("按章节 (共 %d 个)", s.UnitTotal), s.ByUnit, false, false)
	} else {
		top10 := s.ByUnit
		if len(top10) > 10 {
			top10 = top10[:10]
		}
		printSection(fmt.Sprintf("按章节 (Top 10 / 共 %d 个)", s.UnitTotal), top10, false, false)
		if s.UnitTotal > 10 {
			fmt.Printf("  ... 还有 %d 个章节\n", s.UnitTotal-10)
		}
	}

	if s.LowRateCount > 0 {
		fmt.Printf("\n⚠️  正确率 < 50%% 的题目: %d 道\n", s.LowRateCount)
		for _, item := range s.LowRateTop10 {
			fmt.Printf("  [%s] %s...\n", item.Rate, item.Text)
		}
	}

	fmt.Println(strings.Repeat("=", 50))
}

// displayWidth calculates the terminal display width of a string.
func displayWidth(s string) int {
	w := 0
	for _, r := range s {
		if isWide(r) {
			w += 2
		} else {
			w++
		}
	}
	return w
}

func isWide(r rune) bool {
	if !utf8.ValidRune(r) {
		return false
	}
	// CJK characters and fullwidth forms
	if r >= 0x1100 && r <= 0x115F {
		return true
	}
	if r >= 0x2E80 && r <= 0x303E {
		return true
	}
	if r >= 0x3040 && r <= 0x33BF {
		return true
	}
	if r >= 0x3400 && r <= 0x4DBF {
		return true
	}
	if r >= 0x4E00 && r <= 0xA4CF {
		return true
	}
	if r >= 0xA960 && r <= 0xA97C {
		return true
	}
	if r >= 0xAC00 && r <= 0xD7FB {
		return true
	}
	if r >= 0xF900 && r <= 0xFAFF {
		return true
	}
	if r >= 0xFE30 && r <= 0xFE6F {
		return true
	}
	if r >= 0xFF01 && r <= 0xFF60 {
		return true
	}
	if r >= 0xFFE0 && r <= 0xFFE6 {
		return true
	}
	if r >= 0x20000 && r <= 0x2FFFF {
		return true
	}
	if r >= 0x30000 && r <= 0x3FFFF {
		return true
	}
	return false
}

func padRight(s string, width int) string {
	pad := width - displayWidth(s)
	if pad <= 0 {
		return s
	}
	return s + strings.Repeat(" ", pad)
}
